File: pulumi/examples/go-foundation/1-org/iam.go
/*
 * Copyright 2026 Vitruvian Software
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"fmt"

	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/folder"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/organizations"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/projects"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/serviceaccount"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// orgOrFolderIAMMember creates an IAM member at the organization level or at the
// parent-folder level depending on whether a parent_folder is configured.
// This fixes G11: previously the folder branch incorrectly used organizations.NewIAMMember
// with OrgId set to a folder ID. Now it correctly uses folder.NewIAMMember.
func orgOrFolderIAMMember(ctx *pulumi.Context, name string, cfg *OrgConfig, role, member string) error {
	if cfg.ParentFolder == "" {
		if _, err := organizations.NewIAMMember(ctx, name, &organizations.IAMMemberArgs{
			OrgId:  pulumi.String(cfg.OrgID),
			Role:   pulumi.String(role),
			Member: pulumi.String(member),
		}); err != nil {
			return err
		}
	} else {
		if _, err := folder.NewIAMMember(ctx, name+"-folder", &folder.IAMMemberArgs{
			Folder: pulumi.String(cfg.ParentFolder),
			Role:   pulumi.String(role),
			Member: pulumi.String(member),
		}); err != nil {
			return err
		}
	}
	return nil
}

// deployOrgIAM creates all IAM bindings for governance groups on
// org-level projects. This mirrors the Terraform foundation's iam.tf.
func deployOrgIAM(ctx *pulumi.Context, cfg *OrgConfig, proj *OrgProjects) error {
	// ========================================================================
	// 1. Audit Logs Project — IAM for audit_data_users
	// ========================================================================
	if cfg.AuditDataUsers != "" {
		auditGroup := fmt.Sprintf("group:%s", cfg.AuditDataUsers)
		auditRoles := []struct{ name, role string }{
			{"audit-log-viewer", "roles/logging.viewer"},
			{"audit-bq-user", "roles/bigquery.user"},
			{"audit-bq-data-viewer", "roles/bigquery.dataViewer"},
		}
		for _, r := range auditRoles {
			if _, err := projects.NewIAMMember(ctx, r.name, &projects.IAMMemberArgs{
				Project: proj.AuditLogsProjectID,
				Role:    pulumi.String(r.role),
				Member:  pulumi.String(auditGroup),
			}); err != nil {
				return err
			}
		}
	}

	// ========================================================================
	// 2. Billing Export Project — IAM for billing_data_users
	// ========================================================================
	if cfg.BillingDataUsers != "" {
		billingGroup := fmt.Sprintf("group:%s", cfg.BillingDataUsers)

		// Project-level: BQ user + data viewer
		billingRoles := []struct{ name, role string }{
			{"billing-bq-user", "roles/bigquery.user"},
			{"billing-bq-data-viewer", "roles/bigquery.dataViewer"},
		}
		for _, r := range billingRoles {
			if _, err := projects.NewIAMMember(ctx, r.name, &projects.IAMMemberArgs{
				Project: proj.BillingExportProjectID,
				Role:    pulumi.String(r.role),
				Member:  pulumi.String(billingGroup),
			}); err != nil {
				return err
			}
		}

		// Org-level: billing viewer
		if _, err := organizations.NewIAMMember(ctx, "billing-viewer", &organizations.IAMMemberArgs{
			OrgId:  pulumi.String(cfg.OrgID),
			Role:   pulumi.String("roles/billing.viewer"),
			Member: pulumi.String(billingGroup),
		}); err != nil {
			return err
		}
	}

	// ========================================================================
	// 3. Security Reviewer Group — org or folder level (G11 FIX)
	// Previously used organizations.NewIAMMember with OrgId=ParentFolder,
	// which is wrong. Now correctly uses folder.NewIAMMember for folder case.
	// ========================================================================
	if cfg.GCPSecurityReviewer != "" {
		member := fmt.Sprintf("group:%s", cfg.GCPSecurityReviewer)
		if err := orgOrFolderIAMMember(ctx, "security-reviewer", cfg, "roles/iam.securityReviewer", member); err != nil {
			return err
		}
	}

	// ========================================================================
	// 4. Network Viewer Group — org or folder level (G11 FIX)
	// ========================================================================
	if cfg.GCPNetworkViewer != "" {
		member := fmt.Sprintf("group:%s", cfg.GCPNetworkViewer)
		if err := orgOrFolderIAMMember(ctx, "network-viewer", cfg, "roles/compute.networkViewer", member); err != nil {
			return err
		}
	}

	// ========================================================================
	// 5. SCC Admin Group
	// ========================================================================
	if cfg.GCPSCCAdmin != "" {
		member := fmt.Sprintf("group:%s", cfg.GCPSCCAdmin)

		// Org-level: SCC admin editor (only when not under parent_folder)
		if cfg.ParentFolder == "" {
			if _, err := organizations.NewIAMMember(ctx, "org-scc-admin", &organizations.IAMMemberArgs{
				OrgId:  pulumi.String(cfg.OrgID),
				Role:   pulumi.String("roles/securitycenter.adminEditor"),
				Member: pulumi.String(member),
			}); err != nil {
				return err
			}
		}

		// Project-level: SCC admin editor on SCC project (when SCC resources enabled)
		if cfg.EnableSCCResources {
			if _, err := projects.NewIAMMember(ctx, "project-scc-admin", &projects.IAMMemberArgs{
				Project: proj.SCCProjectID,
				Role:    pulumi.String("roles/securitycenter.adminEditor"),
				Member:  pulumi.String(member),
			}); err != nil {
				return err
			}
		}
	}

	// ========================================================================
	// 6. Global Secrets Admin Group
	// ========================================================================
	if cfg.GCPGlobalSecretsAdmin != "" {
		if _, err := projects.NewIAMMember(ctx, "global-secrets-admin", &projects.IAMMemberArgs{
			Project: proj.OrgSecretsProjectID,
			Role:    pulumi.String("roles/secretmanager.admin"),
			Member:  pulumi.String(fmt.Sprintf("group:%s", cfg.GCPGlobalSecretsAdmin)),
		}); err != nil {
			return err
		}
	}

	// ========================================================================
	// 7. KMS Admin Group
	// ========================================================================
	if cfg.GCPKMSAdmin != "" {
		kmsGroup := fmt.Sprintf("group:%s", cfg.GCPKMSAdmin)

		// Project-level: KMS viewer on KMS project
		if _, err := projects.NewIAMMember(ctx, "kms-viewer", &projects.IAMMemberArgs{
			Project: proj.OrgKMSProjectID,
			Role:    pulumi.String("roles/cloudkms.viewer"),
			Member:  pulumi.String(kmsGroup),
		}); err != nil {
			return err
		}

		// Org-level: KMS protected resources viewer (when tracking enabled)
		if cfg.EnableKMSKeyUsageTracking {
			if _, err := organizations.NewIAMMember(ctx, "kms-protected-resources-viewer", &organizations.IAMMemberArgs{
				OrgId:  pulumi.String(cfg.OrgID),
				Role:   pulumi.String("roles/cloudkms.protectedResourcesViewer"),
				Member: pulumi.String(kmsGroup),
			}); err != nil {
				return err
			}
		}
	}

	// ========================================================================
	// 8. KMS Org Service Agent IAM (G3)
	// Grants roles/cloudkms.orgServiceAgent to the KMS service agent at the
	// org level when KMS key usage tracking is enabled. This allows KMS to
	// track key usage across all projects in the organization.
	// Mirrors: google_organization_iam_member "kms_usage_tracking" in iam.tf
	//
	// Gap 4 fix: Pre-create the KMS service agent identity before granting
	// the IAM binding — upstream uses gcloud beta services identity create.
	// Without this, the SA may not exist and the IAM binding references a
	// phantom principal.
	// ========================================================================
	if cfg.EnableKMSKeyUsageTracking {
		// Ensure the KMS organization service agent exists before granting it IAM.
		// This is the Pulumi equivalent of the upstream's:
		//   gcloud beta services identity create --service cloudkms.googleapis.com --organization ${org_id}
		kmsIdentity, err := projects.NewServiceIdentity(ctx, "kms-service-identity", &projects.ServiceIdentityArgs{
			Service: pulumi.String("cloudkms.googleapis.com"),
			Project: proj.OrgKMSProjectID,
		})
		if err != nil {
			return err
		}

		kmsServiceAgent := fmt.Sprintf("serviceAccount:service-org-%s@gcp-sa-cloudkms.iam.gserviceaccount.com", cfg.OrgID)
		if _, err := organizations.NewIAMMember(ctx, "kms-usage-tracking", &organizations.IAMMemberArgs{
			OrgId:  pulumi.String(cfg.OrgID),
			Role:   pulumi.String("roles/cloudkms.orgServiceAgent"),
			Member: pulumi.String(kmsServiceAgent),
		}, pulumi.DependsOn([]pulumi.Resource{kmsIdentity})); err != nil {
			return err
		}
	}

	// ========================================================================
	// 9. Audit Viewer Group IAM (G7)
	// Separate from audit_data_users — this is an optional governance group
	// that gets read access to audit logs and BQ data.
	// Mirrors: google_project_iam_member "audit_log_viewer",
	//          "audit_private_logviewer", "audit_bq_data_viewer" in iam.tf
	// ========================================================================
	if cfg.GCPAuditViewer != "" {
		viewerGroup := fmt.Sprintf("group:%s", cfg.GCPAuditViewer)
		viewerRoles := []struct{ name, role string }{
			{"audit-viewer-log", "roles/logging.viewer"},
			{"audit-viewer-private-log", "roles/logging.privateLogViewer"},
			{"audit-viewer-bq-data", "roles/bigquery.dataViewer"},
		}
		for _, r := range viewerRoles {
			if _, err := projects.NewIAMMember(ctx, r.name, &projects.IAMMemberArgs{
				Project: proj.AuditLogsProjectID,
				Role:    pulumi.String(r.role),
				Member:  pulumi.String(viewerGroup),
			}); err != nil {
				return err
			}
		}
	}

	// ========================================================================
	// 10. Hub-and-Spoke Network SA IAM (G8)
	// When enable_hub_and_spoke is true, grant the networks pipeline SA
	// elevated roles on the hub project to manage compute, SAs, and IAM.
	// Mirrors: google_project_iam_member "network_sa" in projects.tf
	// ========================================================================
	if cfg.EnableHubAndSpoke && cfg.NetworksSAEmail != "" {
		hubAndSpokeRoles := []string{
			"roles/compute.instanceAdmin",
			"roles/iam.serviceAccountAdmin",
			"roles/resourcemanager.projectIamAdmin",
			"roles/iam.serviceAccountUser",
		}
		networkSA := fmt.Sprintf("serviceAccount:%s", cfg.NetworksSAEmail)
		for _, role := range hubAndSpokeRoles {
			if _, err := projects.NewIAMMember(ctx, fmt.Sprintf("net-hub-sa-%s", role), &projects.IAMMemberArgs{
				Project: proj.NetHubProjectID,
				Role:    pulumi.String(role),
				Member:  pulumi.String(networkSA),
			}); err != nil {
				return err
			}
		}
	}

	// ========================================================================
	// 11. CAI Monitoring Builder SA + IAM (G4+G5)
	// Create a dedicated SA for the CAI monitoring Cloud Build pipeline
	// and grant it the roles needed to build and deploy Cloud Functions.
	// Mirrors: sa.tf + iam.tf cai_monitoring_builder in TF foundation
	// ========================================================================
	if cfg.EnableSCCResources {
		if _, err := serviceaccount.NewAccount(ctx, "cai-monitoring-builder", &serviceaccount.AccountArgs{
			Project:     proj.SCCProjectID,
			AccountId:   pulumi.String("cai-monitoring-builder"),
			Description: pulumi.String("Service account for Cloud Build to provision CAI monitoring Cloud Functions"),
		}); err != nil {
			return err
		}

		caiSAMember := proj.SCCProjectID.ApplyT(func(id string) string {
			return fmt.Sprintf("serviceAccount:cai-monitoring-builder@%s.iam.gserviceaccount.com", id)
		}).(pulumi.StringOutput)
		caiRoles := []struct{ name, role string }{
			{"cai-log-writer", "roles/logging.logWriter"},
			{"cai-storage-viewer", "roles/storage.objectViewer"},
			{"cai-ar-writer", "roles/artifactregistry.writer"},
		}
		for _, r := range caiRoles {
			if _, err := projects.NewIAMMember(ctx, r.name, &projects.IAMMemberArgs{
				Project: proj.SCCProjectID,
				Role:    pulumi.String(r.role),
				Member:  caiSAMember,
			}); err != nil {
				return err
			}
		}
	}

	return nil
}

File: pulumi/examples/ts-foundation/1-org/iam.ts
/**
 * Copyright 2026 Vitruvian Software
 *
 * 1-org/iam.ts — IAM bindings for governance groups on org-level projects.
 * Mirrors: 1-org/envs/shared/iam.tf + 1-org/envs/shared/sa.tf in the
 * terraform-example-foundation and 1-org/iam.go in the Go foundation.
 *
 * Creates project-level, org-level, and folder-level IAM bindings that grant
 * the Google Groups (defined in bootstrap) actual permissions on the shared
 * org projects (created in this stage's index.ts).
 */

import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";
import * as command from "@pulumi/command";
import { OrgConfig } from "./config";

export interface OrgProjectRefs {
    auditLogsProjectId: pulumi.Output<string>;
    billingExportProjectId: pulumi.Output<string>;
    orgSecretsProjectId: pulumi.Output<string>;
    orgKmsProjectId: pulumi.Output<string>;
    sccProjectId: pulumi.Output<string>;
    netHubProjectId?: pulumi.Output<string>;
}

/**
 * orgOrFolderIAMMember creates an IAM member at the org level or
 * at the parent-folder level depending on configuration.
 * Matches Go foundation's orgOrFolderIAMMember helper.
 */
function orgOrFolderIAMMember(
    name: string,
    cfg: OrgConfig,
    role: string,
    member: string,
): void {
    if (cfg.parentFolder === "") {
        new gcp.organizations.IAMMember(name, {
            orgId: cfg.orgId,
            role,
            member,
        });
    } else {
        new gcp.folder.IAMMember(`${name}-folder`, {
            folder: cfg.parentFolder,
            role,
            member,
        });
    }
}

/**
 * deployOrgIAM creates all IAM bindings for governance groups.
 * Port of Go foundation's deployOrgIAM (iam.go:56-319).
 */
export function deployOrgIAM(cfg: OrgConfig, proj: OrgProjectRefs): void {
    // ========================================================================
    // 1. Audit Logs Project — IAM for audit_data_users
    // ========================================================================
    const auditDataUsers = cfg.requiredGroups.audit_data_users;
    if (auditDataUsers) {
        const auditGroup = `group:${auditDataUsers}`;
        const auditRoles: [string, string][] = [
            ["audit-log-viewer", "roles/logging.viewer"],
            ["audit-bq-user", "roles/bigquery.user"],
            ["audit-bq-data-viewer", "roles/bigquery.dataViewer"],
        ];
        for (const [name, role] of auditRoles) {
            new gcp.projects.IAMMember(name, {
                project: proj.auditLogsProjectId,
                role,
                member: auditGroup,
            });
        }
    }

    // ========================================================================
    // 2. Billing Export Project — IAM for billing_data_users
    // ========================================================================
    const billingDataUsers = cfg.requiredGroups.billing_data_users;
    if (billingDataUsers) {
        const billingGroup = `group:${billingDataUsers}`;

        // Project-level: BQ user + data viewer
        const billingRoles: [string, string][] = [
            ["billing-bq-user", "roles/bigquery.user"],
            ["billing-bq-data-viewer", "roles/bigquery.dataViewer"],
        ];
        for (const [name, role] of billingRoles) {
            new gcp.projects.IAMMember(name, {
                project: proj.billingExportProjectId,
                role,
                member: billingGroup,
            });
        }

        // Org-level: billing viewer
        new gcp.organizations.IAMMember("billing-viewer", {
            orgId: cfg.orgId,
            role: "roles/billing.viewer",
            member: billingGroup,
        });
    }

    // ========================================================================
    // 3. Security Reviewer Group — org or folder level
    // ========================================================================
    const securityReviewer = cfg.gcpGroups.securityReviewer;
    if (securityReviewer) {
        orgOrFolderIAMMember("security-reviewer", cfg, "roles/iam.securityReviewer", `group:${securityReviewer}`);
    }

    // ========================================================================
    // 4. Network Viewer Group — org or folder level
    // ========================================================================
    const networkViewer = cfg.gcpGroups.networkViewer;
    if (networkViewer) {
        orgOrFolderIAMMember("network-viewer", cfg, "roles/compute.networkViewer", `group:${networkViewer}`);
    }

    // ========================================================================
    // 5. SCC Admin Group
    // ========================================================================
    const sccAdmin = cfg.gcpGroups.sccAdmin;
    if (sccAdmin) {
        const member = `group:${sccAdmin}`;

        // Org-level: SCC admin editor (only when not under parent_folder)
        if (cfg.parentFolder === "") {
            new gcp.organizations.IAMMember("org-scc-admin", {
                orgId: cfg.orgId,
                role: "roles/securitycenter.adminEditor",
                member,
            });
        }

        // Project-level: SCC admin editor on SCC project
        new gcp.projects.IAMMember("project-scc-admin", {
            project: proj.sccProjectId,
            role: "roles/securitycenter.adminEditor",
            member,
        });
    }

    // ========================================================================
    // 6. Global Secrets Admin Group
    // ========================================================================
    const globalSecretsAdmin = cfg.gcpGroups.globalSecretsAdmin;
    if (globalSecretsAdmin) {
        new gcp.projects.IAMMember("global-secrets-admin", {
            project: proj.orgSecretsProjectId,
            role: "roles/secretmanager.admin",
            member: `group:${globalSecretsAdmin}`,
        });
    }

    // ========================================================================
    // 7. KMS Admin Group
    // ========================================================================
    const kmsAdmin = cfg.gcpGroups.kmsAdmin;
    if (kmsAdmin) {
        const kmsGroup = `group:${kmsAdmin}`;

        // Project-level: KMS viewer on KMS project
        new gcp.projects.IAMMember("kms-viewer", {
            project: proj.orgKmsProjectId,
            role: "roles/cloudkms.viewer",
            member: kmsGroup,
        });
    }

    // ========================================================================
    // 7b. KMS Protected Resources Viewer (mirrors TF iam.tf:174-179)
    // ========================================================================
    if (kmsAdmin && cfg.enableKmsKeyUsageTracking) {
        const kmsGroup = `group:${kmsAdmin}`;
        new gcp.organizations.IAMMember("kms-protected-resources-viewer", {
            orgId: cfg.orgId,
            role: "roles/cloudkms.protectedResourcesViewer",
            member: kmsGroup,
        });
    }

    // ========================================================================
    // 8. KMS Org Service Agent IAM
    // Grants roles/cloudkms.orgServiceAgent to the KMS service agent.
    // Upstream TF uses `gcloud beta services identity create --organization`
    // to provision the org-level KMS service agent. The Pulumi
    // gcp.projects.ServiceIdentity only creates a project-level identity,
    // so we use a Command resource for the org-level agent.
    // Gated behind enableKmsKeyUsageTracking (default: true in TF).
    // ========================================================================
    if (cfg.enableKmsKeyUsageTracking) {
        const kmsOrgIdentity = new command.local.Command("kms-org-service-identity", {
            create: `gcloud beta services identity create --service cloudkms.googleapis.com --organization ${cfg.orgId}`,
        });

        const kmsServiceAgent = `serviceAccount:service-org-${cfg.orgId}@gcp-sa-cloudkms.iam.gserviceaccount.com`;
        new gcp.organizations.IAMMember("kms-usage-tracking", {
            orgId: cfg.orgId,
            role: "roles/cloudkms.orgServiceAgent",
            member: kmsServiceAgent,
        }, { dependsOn: [kmsOrgIdentity] });
    }

    // ========================================================================
    // 9. Audit Viewer Group IAM
    // Separate from audit_data_users — optional governance group.
    // ========================================================================
    const auditViewer = cfg.gcpGroups.auditViewer;
    if (auditViewer) {
        const viewerGroup = `group:${auditViewer}`;
        const viewerRoles: [string, string][] = [
            ["audit-viewer-log", "roles/logging.viewer"],
            ["audit-viewer-private-log", "roles/logging.privateLogViewer"],
            ["audit-viewer-bq-data", "roles/bigquery.dataViewer"],
        ];
        for (const [name, role] of viewerRoles) {
            new gcp.projects.IAMMember(name, {
                project: proj.auditLogsProjectId,
                role,
                member: viewerGroup,
            });
        }
    }

    // ========================================================================
    // 10. Hub-and-Spoke Network SA IAM
    // When enable_hub_and_spoke is true, grant the networks pipeline SA
    // elevated roles on the hub project.
    // ========================================================================
    if (cfg.enableHubAndSpoke && cfg.networksStepTerraformSaEmail && proj.netHubProjectId) {
        const hubAndSpokeRoles = [
            "roles/compute.instanceAdmin",
            "roles/iam.serviceAccountAdmin",
            "roles/resourcemanager.projectIamAdmin",
            "roles/iam.serviceAccountUser",
        ];
        const networkSA = `serviceAccount:${cfg.networksStepTerraformSaEmail}`;
        for (const role of hubAndSpokeRoles) {
            new gcp.projects.IAMMember(`net-hub-sa-${role.split("/")[1]}`, {
                project: proj.netHubProjectId,
                role,
                member: networkSA,
            });
        }
    }

    // ========================================================================
    // 11. CAI Monitoring Builder SA + IAM
    // Create dedicated SA for the CAI monitoring Cloud Build pipeline.
    // Gated behind enableSccResources (matching upstream TF).
    // ========================================================================
    if (cfg.enableSccResources) {
        const caiBuilderSa = new gcp.serviceaccount.Account("cai-monitoring-builder", {
            project: proj.sccProjectId,
            accountId: "cai-monitoring-builder",
            description: "Service account for Cloud Build to provision CAI monitoring Cloud Functions",
            createIgnoreAlreadyExists: true,
        });

        const caiSAMember = pulumi.interpolate`serviceAccount:${caiBuilderSa.email}`;
        const caiRoles: [string, string][] = [
            ["cai-log-writer", "roles/logging.logWriter"],
            ["cai-storage-viewer", "roles/storage.objectViewer"],
            ["cai-ar-writer", "roles/artifactregistry.writer"],
        ];
        for (const [name, role] of caiRoles) {
            new gcp.projects.IAMMember(name, {
                project: proj.sccProjectId,
                role,
                member: caiSAMember,
            });
        }
    }
}
